include/deal.II/base/tensor_polynomials_base.h

```diff
@@ -0,0 +1,181 @@
+ * @p FE_Poly and @p FE_PolyTensor. An object of this type is stored
+ * as a member variable in the finite element class.
+ *
+ * <h3>Deriving classes</h3>
+ *
+ * Any derived class must provide the most basic properties for shape
+ * functions evaluated on the reference cell. This includes, but is not
+ * limited to, implementing the <tt>compute</tt>, <tt>name</tt>, and
+ * <tt>clone</tt> member functions. These functions are necessary to store the
+ * most basic information of how the polynomials in the derived class evaluate
+ * at a given point on the reference cell. More information on each function can
+ * be found in the corresponding function's documentation.
+ *
+ * Some classes that derive from this class include
+ * <ul>
+ *   <li> <tt>PolynomialsABF</tt>
+ *   <li> <tt>PolynomialsBDM</tt>
+ *   <li> <tt>PolynomialsBernardiRaugel</tt>
+ *   <li> <tt>PolynomialsNedelec</tt>
+ *   <li> <tt>PolynomialsRaviartThomas</tt>
+ *   <li> <tt>PolynomialsRT_Bubbles</tt>
+ * </ul>
+ *
+ * @ingroup Polynomials
+ * @author Graham Harper
+ * @date 2019
+ */
+template <int dim>
+class TensorPolynomialsBase
+{
+public:
+  /**
+   * Constructor. This takes a integer @p k from the finite element class,
+   * which assigned to @p my_degree on construction. @p n is assigned to @p n_pols
+   * on construction.
+   */
+  TensorPolynomialsBase(const unsigned int k, const unsigned int n);
+
+  /**
+   * Move constructor.
+   */
+  TensorPolynomialsBase(TensorPolynomialsBase<dim> &&) = default; // NOLINT
+
+  /**
+   * Copy constructor.
+   */
+  TensorPolynomialsBase(const TensorPolynomialsBase<dim> &) = default;
+
+  /**
+   * Virtual destructor. Makes sure that pointers to this class are deleted
+   * properly.
+   */
+  virtual ~TensorPolynomialsBase() = default;
+
+  /**
+   * Compute the value and the derivatives of the polynomials at
+   * @p unit_point.
+   *
+   * The size of the vectors must either be zero or equal <tt>n()</tt>.  In
+   * the first case, the function will not compute these values.
+   *
+   * If you need values or derivatives of all polynomials then use this
+   * function, rather than using any of the <tt>compute_value</tt>,
+   * <tt>compute_grad</tt> or <tt>compute_grad_grad</tt> functions, see below,
+   * in a loop over all tensor product polynomials.
+   */
+  virtual void
+  compute(const Point<dim> &           unit_point,
+          std::vector<Tensor<1, dim>> &values,
+          std::vector<Tensor<2, dim>> &grads,
+          std::vector<Tensor<3, dim>> &grad_grads,
+          std::vector<Tensor<4, dim>> &third_derivatives,
+          std::vector<Tensor<5, dim>> &fourth_derivatives) const = 0;
+
+  /**
+   * Return the number of polynomials.
+   */
+  unsigned int
+  n() const;
+
+  /**
+   * Return the highest polynomial degree of polynomials represented by this
+   * class. A derived class may override this if its value is different from
+   * @p my_degree.
+   */
+  virtual unsigned int
+  degree() const;
```

Comment:
I think you should either:
* Pass the degree to the constrtuctor and return it here via a non-`virtual` function
* Make the function `=0` and let derived classes implement it.

Having both is not useful: The value passed to the constructor does, after all, come from the derived class. The derived class *knows* what the proper value is, so there is no reason for the constructor of the derived class to pass something down to the constructor of the base class, but then later return something *different* in its own implementation of the virtual function.

My preference would be the first option above.

Reply:
My worry is based on how different the implementations of the `degree()` function is for each polynomial space. For example, looking at BDM, BernardiRaugel, and RaviartThomas shows each is implemented totally differently.
```
PolynomialsBDM<dim>::degree() const
{
  return polynomial_space.degree() - 1;
}
```

```
PolynomialsBernardiRaugel<dim>::degree() const
{
  return 2;
}
```

```
PolynomialsRaviartThomas<dim>::degree() const
{
  return my_degree;
}
```
My belief is they should be virtual so that the base class knows they are there, but not abstract since the base implementation shouldn't always be overridden.

Or, should the class contain two member variables? One for the degree of the space (the 0 in RT_0) and one for the polynomial degree of the space (1 for RT_0), and then `degree()` always returns the polynomial degree? I apologize if I'm making this more complicated than necessary, but I just want to make sure it's compatible with all the classes that will derive from it.

Reply:
These are all values known at the time the constructor is called in the derived class. These derived class constructors may as well pass this value down to the constructor of the base class:
```
  PolynomialsBDM::PolynomialsBDG (const uint degree) :
    TensorPolynomials (...something that can be computed from degree..., ...)
  {}
```

Reply:
And even simpler in the other two cases!

Reply:
I see. That works! So this would change the implementations but keep the base class simple.